Common/Orders/Fills/EquityFillModel.cs

```diff
@@ -391,9 +391,21 @@ public virtual OrderEvent MarketOnOpenFill(Security asset, MarketOnOpenOrder ord
 
             // wait until market open
             // make sure the exchange is open/normal market hours before filling
-            if (!IsExchangeOpen(asset, false)) return fill;
+            if (!IsExchangeOpen(asset, false))
+            {
+                return fill;
+            }
 
-            fill.FillPrice = GetPricesCheckingPythonWrapper(asset, order.Direction).Open;
+            // For non-daily resolution, check whether the last trade bar
+            // If the first data after the market opens is a quote bar, the open of the last trade bar is not the open
+            // for the current day, but the last open in cache
+            var resolution = Parameters.ConfigProvider.GetSubscriptionDataConfigs(asset.Symbol).GetHighestResolution();
+            if (resolution != Resolution.Daily && !asset.Exchange.DateTimeIsOpen(lastTradeBar.EndTime))
```

Comment:
`asset.Exchange.DateTimeIsOpen(lastTradeBar.EndTime)` will be false of the bar ending at the market close, which I think should be a valid bar to use? maybe we could just compare the lastTradeBar date > order date?